explorer/src/Pos/Explorer/ExtraContext.hs

```diff
@@ -37,8 +86,91 @@ instance Monad m => HasGenesisRedeemAddressInfo (ExtraContextT m) where
+-------------------------------------------------------------------------------------
+-- Explorer mock mode
+--
+-- The simple data structure that encapsulates functions that use CSL. We want to "cut"
+-- them out of the picture in order to be able to mock them.
+-------------------------------------------------------------------------------------
+
+data ExplorerMockMode = ExplorerMockMode
+    { emmGetTipBlock
+          :: forall m. MonadBlockDB m => m Block
+    , emmGetPageBlocks
+          :: forall m. MonadDBRead m => Page -> m (Maybe [HeaderHash])
+    , emmGetBlundFromHH
+          :: forall m. MonadBlockDB m => HeaderHash -> m (Maybe Blund)
+    , emmGetSlotStart
+          :: forall ctx m. MonadSlotsData ctx m => SlotId -> m (Maybe Timestamp)
+    , emmGetLeadersFromEpoch
+          :: forall m. MonadDBRead m => EpochIndex -> m (Maybe SlotLeaders)
+    }
+
+-- | This is what we use in production when we run Explorer.
+prodMode :: ExplorerMockMode
+prodMode = ExplorerMockMode {
+      emmGetTipBlock            = getTipBlock,
+      emmGetPageBlocks          = getPageBlocks,
+      emmGetBlundFromHH         = blkGetBlund,
+      emmGetSlotStart           = getSlotStart,
+      emmGetLeadersFromEpoch    = getLeaders
+    }
+
+-- | So we can just reuse the default instance and change individial functions.
+-- On one side, it removes the compile error(s) for having all functions implemented.
+-- On the other side, it moves that error into runtime and enables simple mocking.
+-- This is a good thing once we have a larger amount of functions, like in _explorer_,
+-- and this gives us the flexibility to "mock" whichever we want.
+instance Default (ExplorerMockMode) where
+  def = ExplorerMockMode {
+        emmGetTipBlock            = errorImpl,
+        emmGetPageBlocks          = errorImpl,
+        emmGetBlundFromHH         = errorImpl,
+        emmGetSlotStart           = errorImpl,
+        emmGetLeadersFromEpoch    = errorImpl
+      }
+    where
+      errorImpl = error "Cannot be used, please implement this function!"
+
+
+-------------------------------------------------------------------------------------
+-- Explorer interface instance
+-------------------------------------------------------------------------------------
+
+-- | We use this for an external CSL functions representation so we can mock them when
+-- testing.
+class HasExplorerCSLInterface m where
+    getTipBlockCSLI :: m Block
+    getPageBlocksCSLI :: Page -> m (Maybe [HeaderHash])
+    getBlundFromHHCSLI :: HeaderHash -> m (Maybe Blund)
+    getSlotStartCSLI :: SlotId -> m (Maybe Timestamp)
+    getLeadersFromEpochCSLI :: EpochIndex -> m (Maybe SlotLeaders)
+
+-- | The instance for external CSL functions.
+instance (Monad m, MonadBlockDB m, MonadDBRead m, MonadSlotsData ctx m) =>
+    HasExplorerCSLInterface (ExtraContextT m) where
+
+    getTipBlockCSLI = do
+        extraCtx <- Ether.ask @ExtraContext
+        let explorerMockMode = ecExplorerMockMode extraCtx
```

Comment:
Maybe we could extract these two lines above to get `ExplorerMockMode` into a function, because they are used in all following functions `get...`. Just to be more DRY.

Reply:
We could do that, but it's just two lines and it's pretty simple to understand as-is. If we add another function things will become a bit more unreadable.
But yes, good catch, I was considering it :smile: